drivers/i2s/i2s_nrfx.c

```diff
@@ -0,0 +1,788 @@
+		nrfx_i2s_buffers_t next = { 0 };
+
+		if (drv_data->active_dir != I2S_DIR_RX) { /* -> TX active */
+			if (drv_data->stop) {
+				/* If the stream is to be stopped, don't get
+				 * the next TX buffer from the queue, instead
+				 * supply the one used last time (it won't be
+				 * transferred, the stream will stop right
+				 * before this buffer would be started again).
+				 */
+				next.p_tx_buffer = drv_data->last_tx_buffer;
+			} else if (get_next_tx_buffer(drv_data, &next)) {
+				/* Each time a new TX buffer is retrieved from
+				 * the queue, update the corresponding pointer.
+				 */
+				drv_data->last_tx_buffer = next.p_tx_buffer;
+			} else if (drv_data->state == I2S_STATE_STOPPING) {
+				/* If there are no more TX blocks queued and
+				 * the current state is STOPPING (so the DRAIN
+				 * command was triggered) it is time to finish
+				 * the transfer.
+				 */
+				drv_data->stop = true;
+				/* Supply the same buffer as last time, it won't
+				 * be tranferred anyway, as the transfer will
+				 * stop earlier.
+				 */
+				next.p_tx_buffer = drv_data->last_tx_buffer;
+			} else {
+				LOG_ERR("No TX buffer available");
+				nrfx_i2s_stop();
+				drv_data->state = I2S_STATE_ERROR;
+				return;
+			}
+		}
+
+		if (drv_data->active_dir != I2S_DIR_TX) { /* -> RX active */
+			if (!get_next_rx_buffer(drv_data, &next)) {
+				nrfx_i2s_stop();
+				drv_data->state = I2S_STATE_ERROR;
+				return;
+			}
+		}
+
+		LOG_DBG("Buffers: %p/%p", next.p_tx_buffer, next.p_rx_buffer);
+		nrfx_i2s_next_buffers_set(&next);
+	}
+}
+
+static void purge_queue(struct i2s_nrfx_drv_data *drv_data,
+			enum i2s_dir dir)
+{
+	void *mem_block;
+
+	if (dir == I2S_DIR_TX || dir == I2S_DIR_BOTH) {
+		while (1) {
+			mem_block = queue_get(&drv_data->tx.queue);
+			if (mem_block) {
+				k_mem_slab_free(drv_data->tx.config.mem_slab,
+						&mem_block);
+				k_sem_give(&drv_data->tx.sem);
+			} else {
+				break;
+			}
+		}
+	}
+
+	if (dir == I2S_DIR_RX || dir == I2S_DIR_BOTH) {
+		while (k_sem_take(&drv_data->rx.sem, K_NO_WAIT) == 0) {
+			mem_block = queue_get(&drv_data->rx.queue);
+			if (mem_block) {
+				k_mem_slab_free(drv_data->rx.config.mem_slab,
+						&mem_block);
+			}
+		}
+	}
+}
+
+static int i2s_nrfx_configure(const struct device *dev, enum i2s_dir dir,
+			      const struct i2s_config *i2s_cfg)
+{
+	struct i2s_nrfx_drv_data *drv_data = dev->data;
+	const struct i2s_nrfx_drv_cfg *drv_cfg = dev->config;
+	nrfx_i2s_config_t nrfx_cfg;
+
+	if (drv_data->state != I2S_STATE_READY) {
+		LOG_ERR("Cannot configure in state: %d", drv_data->state);
+		return -EINVAL;
+	}
+
+	if (i2s_cfg->frame_clk_freq == 0) { /* -> reset state */
+		purge_queue(drv_data, dir);
+		if (dir == I2S_DIR_TX || I2S_DIR_BOTH) {
+			drv_data->tx_configured = false;
+			memset(&drv_data->tx.nrfx_cfg, 0,
+			       sizeof(drv_data->tx.nrfx_cfg));
+			memset(&drv_data->tx.config, 0,
+			       sizeof(drv_data->tx.config));
+		}
+		if (dir == I2S_DIR_RX || I2S_DIR_BOTH) {
```

Comment:
Shouldn't these be `if (dir == I2S_DIR_TX || dir == I2S_DIR_BOTH)`  and `if (dir == I2S_DIR_RX || dir == I2S_DIR_BOTH)`?

Reply:
Thanks for spotting that. Fixed.